design-documents/storefront/catalog/product-options-and-variants.md

```diff
@@ -0,0 +1,363 @@
+   
+### Actual usages
+
+The application distinguishes two approaches to manage options:
+
+**Approach 1, "One to many"**: Multiple options selections lead a shopper to a selection of the single variation.
+
+Example: configurable products
+![](https://app.lucidchart.com/publicSegments/view/26cd0b67-13c1-44e4-8b61-cada36c67010/image.png)
+
+**Approach 2, "One to one"**: The selection of multiple options leads to multiple product variants.
+
+Examples: bundle product, customizable product, downloadable products.
+![](https://app.lucidchart.com/publicSegments/view/fea3f950-6f2f-4e46-90e4-f8ee4b6877f7/image.png)
+Both of approaches could be used together.
+
+Example: configurable product with customizable option.
+
+
+## Competitors analysis
+During my research, I analyzed documentation on how options are covered by the competitors.
+commercetools and Shopify attack this domain with product variants that represent all the possible options.
+Despite the domain simplification by reducing the number of entities both systems constrained with the size of a product which makes it impossible to compose complex business cases that we frequently may face in the Magento ecosystem.
+
+[Commercetools: Modeling Products](https://docs.commercetools.com/product-modeling-products)
+
+[Shopify: Variants](https://help.shopify.com/en/manual/products/variants)
+
+Bigcommerce has a more complex option representation then Shopify or Commercetools.
+in some way, I would compare their options features with Magento 2 x functionalities.
+Bigcommerce segregates options from variants and allows to associate intersection of options with a new product, and as a result, I believe, struggles from the same disease as Magento does - redundancy in the variant data.
+
+[Bigcommerce: Product Options](https://support.bigcommerce.com/s/article/Options-SKUs-Rules)
+
+## Goals
+
+* Eliminate hard dependency on prices from options to make possible use promotions and B2B prices for customizable options, bundle, and downloadable products.
+* Eliminate variants redundancy
+    * Catalog should not have a variant per each option values combination, only if necessary.
+    * Catalog should not create a product per variant if this product never used.
+* Hide the variant matrix from storefront client, use it only in the context of operations.
+* Align a similar option with a single data structure to manage them together.
+
+
+## Modeling Options and Variants Data Objects
+
+The great advantage of Magento 2 - modularity was not a real thing for the product options during Magento the lifetime.
+Since times of Magento 1.x option prices coupled into options, inventory records ignored even options with assigned SKUs.
+The only thing that had to play it together "Bundle" product was designed too complex to treat it as universal solution it had to become.
```

Comment:
Check attribute references from commercetools for bundling approach. In that approach an attribute can have a reference type, which can be used for bundling. Meaning you can create a Product Type, with reference type attributes of other product type.

Reply:
that's exactly what we have with configurable products, it has a perfect sense for admin functionality to enforce data consistency but in fact redundant for the storefront.

Reply:
what I am trying to say, that catalog is a projection od data that's why we probably do not want to integrate back-office constants. especially taking to account mutable sources of products that could be connected.

Reply:
I believe the concept is applied on product catalog modeling, whether storefront or admin; the same concept can be applied.